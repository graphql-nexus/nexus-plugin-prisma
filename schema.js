module.exports = require('./dist/schema-entrypoint')
