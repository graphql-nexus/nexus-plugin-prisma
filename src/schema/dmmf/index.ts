export * from './utils'
export * from './transformer'
export * from './DmmfDocument'
export * from './DmmfTypes'
