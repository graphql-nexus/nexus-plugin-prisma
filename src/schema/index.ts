import { plugin } from '@nexus/schema'
import { build as buildNexusPrismaTypes, Options, InternalPublicOptions } from './builder'
import { CRUD_NOT_ENABLED } from './warnings'

/**
 * Create a nexus-plugin-prisma instance to be passed into the Nexus plugins array.
 *
 * The nexus plugin extends the Nexus DSL with `t.model` and `t.crud`. These
 * allow you to expose data from your Prisma layer and operations against that
 * data. Once this plugin is installed, you can do things such as:
 *
 * ```ts
 * objectType({
 *   name: 'User',
 *   definition(t) {
 *     t.model.id()
 *     t.model.email()
 *   }
 * })
 *
 * queryType({
 *   definition (t) {
 *      t.crud.user()
 *      t.crud.users({ filtering: true, ordering: true })
 *    }
 * })
 * ```
 *
 * You must ensure the Prisma Client JS has been generated prior as it provides a
 * data representation of the available models and CRUD operations against them.
 *
 * Like Nexus, NexusPrisma has its own Typegen. By default It will be generated
 * synchronously when Nexus runs the plugin, _if_ `process.env.NODE_ENV` is
 * `undefined` or `"development"`. Typegen can be explicitly enabled or disabled
 * via the shouldGenerateArtifacts option. This mirrors Nexus' own typegen
 * approach. This system will change once the Nexus plugin system has first
 * class support for typegen.
 */
export function nexusSchemaPrisma(options?: Options) {
  const allOptions: InternalPublicOptions = options ?? {}
  let wasCrudUsedButDisabled: null | (() => boolean) = null

  return plugin({
    name: 'nexus-plugin-prisma',
    onInstall: (nexusBuilder) => {
      const { types, wasCrudUsedButDisabled: wasCrudUsed } = buildNexusPrismaTypes({
        ...allOptions,
        nexusBuilder,
      })

      wasCrudUsedButDisabled = wasCrudUsed

      return {
        types,
      }
    },
    onBeforeBuild() {
      if (wasCrudUsedButDisabled?.() === true) {
        if (allOptions.initializedByFramework === true) {
          CRUD_NOT_ENABLED.framework()
        } else {
          CRUD_NOT_ENABLED.schema()
        }
      }
    },
  })
}

export { Options } from './builder'
