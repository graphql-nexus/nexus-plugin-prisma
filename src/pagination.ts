import { DMMF } from '@prisma/client/runtime'
import { DmmfTypes, getReturnTypeName } from './dmmf'

interface PaginationResult {
  cursor?: object
  skip?: string | number
  take?: string | number
  orderBy?: { [x: string]: 'asc' | 'desc' }
  [x: string]: any
}

export interface PaginationStrategy<T = object> {
  paginationArgNames: string[]
  transformDmmfArgs(params: {
    args: DmmfTypes.SchemaArg[]
    paginationArgNames: string[]
    field: DMMF.SchemaField
  }): DmmfTypes.SchemaArg[]
  resolve(args: T): PaginationResult
}

const relayLikePaginationArgs: Record<
  'first' | 'last' | 'before' | 'after',
  (typeName: string) => DmmfTypes.SchemaArg
> = {
  first: () => ({
    name: 'first',
    inputType: {
      type: 'Int',
      kind: 'scalar',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
  last: () => ({
    name: 'last',
    inputType: {
      type: 'Int',
      kind: 'scalar',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
  before: (typeName: string) => ({
    name: 'before',
    inputType: {
      type: `${typeName}WhereUniqueInput`,
      kind: 'object',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
  after: (typeName: string) => ({
    name: 'after',
    inputType: {
      type: `${typeName}WhereUniqueInput`,
      kind: 'object',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
}

interface RelayLikePaginationArgs {
  first?: number
  last?: number
  before?: object
  after?: object
}

interface PaginationStrategies {
  relay: PaginationStrategy<RelayLikePaginationArgs>
  prisma: PaginationStrategy<PrismaPaginationArgs>
}

const prismaPaginationArgs: Record<'take' | 'skip' | 'cursor', (typeName: string) => DmmfTypes.SchemaArg> = {
  take: () => ({
    name: 'take',
    inputType: {
      type: 'Int',
      kind: 'scalar',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
  skip: () => ({
    name: 'skip',
    inputType: {
      type: 'Int',
      kind: 'scalar',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
  cursor: (typeName: string) => ({
    name: 'cursor',
    inputType: {
      type: `${typeName}WhereUniqueInput`,
      kind: 'object',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
}

interface PrismaPaginationArgs {
  cursor?: object
  take?: number
  skip?: number
}

const PaginationStrategies: PaginationStrategies = {
  relay: {
    paginationArgNames: Object.keys(relayLikePaginationArgs),
    transformDmmfArgs({ args, paginationArgNames, field }) {
      const fieldOutputTypeName = getReturnTypeName(field.outputType.type)

      // Remove old pagination args
      args = args.filter((a) => !paginationArgNames.includes(a.name))

      // Push new pagination args
      args.push(
        relayLikePaginationArgs.first(fieldOutputTypeName),
        relayLikePaginationArgs.last(fieldOutputTypeName),
        relayLikePaginationArgs.before(fieldOutputTypeName),
        relayLikePaginationArgs.after(fieldOutputTypeName)
      )

      return args
    },
    resolve(args) {
      const { first, last, before, after } = args

      // If no pagination set, don't touch the args
      if (!first && !last && !before && !after) {
        return args
      }

      /**
       * This is currently only possible with js transformation on the result. eg:
       * after: 1, last: 1
       * ({
       *   cursor: { id: $before },
       *   take: Number.MAX_SAFE_INTEGER,
       *   skip: 1
       * }).slice(length - $last, length)
       */
      if (after && last) {
        throw new Error(`after and last can't be set simultaneously`)
      }

      /**
       * This is currently only possible with js transformation on the result. eg:
       * before: 4, first: 1
       * ({
       *   cursor: { id: $before },
       *   take: Number.MIN_SAFE_INTEGER,
       *   skip: 1
       * }).slice(0, $first)
       */
      if (before && first) {
        throw new Error(`before and first can't be set simultaneously`)
      }

      // Edge-case: simulates a single `before` with a hack
      if (before && !first && !last && !after) {
        return {
          cursor: before,
          skip: 1,
          take: Number.MIN_SAFE_INTEGER,
        }
      }

      const take = resolveTake(first, last, before)
      const cursor = resolveCursor(before, after)
      const skip = resolveSkip(cursor)

      delete args.first
      delete args.last
      delete args.before
      delete args.after

      const newArgs = {
        take,
        cursor,
        skip,
        ...args,
      }

      return newArgs
    },
  },
  prisma: {
    paginationArgNames: Object.keys(prismaPaginationArgs),
    transformDmmfArgs({ args, paginationArgNames, field }) {
      return args
    },
    resolve(args) {
      return args
    },
  },
}

function resolveTake(
  first: number | undefined,
  last: number | undefined,
  before: object | undefined
): number | undefined {
  if (first && last) {
    throw new Error(`first and last can't be set simultaneously`)
  }

  if (first) {
    if (first < 0) {
      throw new Error(`first can't be negative`)
    }
    return first
  }

  if (last) {
    if (last < 0) {
      throw new Error(`last can't be negative`)
    }

    if (last === 0) {
      return 0
    }

    return last * -1
  }

  return undefined
}

function resolveCursor(before: object | undefined, after: object | undefined) {
  if (before && after) {
    throw new Error(`before and after can't be set simultaneously`)
  }

  return before ?? after
}

function resolveSkip(cursor: object | undefined) {
  if (cursor) {
    return 1
  }

  return undefined
}

export default PaginationStrategies
