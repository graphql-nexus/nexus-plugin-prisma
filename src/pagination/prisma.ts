import { PaginationStrategy } from '.'
import { Helpers as DMMFHelpers, InternalDMMF } from '../dmmf'
import { keys } from '../utils'

interface PrismaPaginationArgs {
  cursor?: object
  take?: number
  skip?: number
}

const prismaPaginationArgsToDmmfArgs: Record<
  'take' | 'skip' | 'cursor',
  (typeName: string) => InternalDMMF.SchemaArg
> = {
  take: () => ({
    name: 'take',
    inputType: {
      type: 'Int',
      kind: 'scalar',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
  skip: () => ({
    name: 'skip',
    inputType: {
      type: 'Int',
      kind: 'scalar',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
  cursor: (typeName: string) => ({
    name: 'cursor',
    inputType: {
      type: `${typeName}WhereUniqueInput`,
      kind: 'object',
      isRequired: false,
      isList: false,
      isNullable: false,
    },
  }),
}

export const prismaStrategy: PaginationStrategy<PrismaPaginationArgs> = {
  paginationArgNames: keys(prismaPaginationArgsToDmmfArgs),
  transformDmmfArgs({ paginationArgNames, args, field }) {
    // todo why not using internal dmmf here?
    const fieldOutputTypeName = DMMFHelpers.getTypeName(field.outputType.type)

    // Remove old pagination args
    args = args.filter((dmmfArg) => !paginationArgNames.includes(dmmfArg.name))

    // Push new pagination args
    args.push(
      prismaPaginationArgsToDmmfArgs.take(fieldOutputTypeName),
      prismaPaginationArgsToDmmfArgs.skip(fieldOutputTypeName),
      prismaPaginationArgsToDmmfArgs.cursor(fieldOutputTypeName)
    )

    return args
  },
  resolve(args) {
    return args
  },
}
