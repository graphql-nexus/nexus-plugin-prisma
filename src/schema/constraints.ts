import { DmmfDocument } from './dmmf'

/**
 * Find the unique identifiers necessary to indentify a field
 *
 * Unique fields for a model can be one of (in this order):
 * 1. One (and only one) field with an @id annotation
 * 2. Multiple fields with @@id clause
 * 3. One (and only one) field with a @unique annotation (if there are multiple, use the first one)
 * 4. Multiple fields with a @@unique clause
 */
export function resolveUniqueIdentifiers(typeName: string, dmmf: DmmfDocument): string[] {
  const model = dmmf.getModelOrThrow(typeName)

  // Try finding 1.
  const singleIdField = model.fields.find((f) => f.isId)

  if (singleIdField) {
    return [singleIdField.name]
  }

  // Try finding 2.
  if (model.idFields && model.idFields.length > 0) {
    return model.idFields
  }

  const singleUniqueField = model.fields.find((f) => f.isUnique)

  if (singleUniqueField) {
    return [singleUniqueField.name]
  }

  if (model.uniqueFields && model.uniqueFields.length > 0) {
    return model.uniqueFields[0]
  }

  throw new Error(`Unable to resolve a unique identifier for the Prisma model: ${model.name}`)
}

export function findMissingUniqueIdentifiers(
  data: Record<string, any>,
  uniqueIdentifiers: string[]
): string[] | null {
  const missingIdentifiers: string[] = []

  for (const identifier of uniqueIdentifiers) {
    if (!data[identifier]) {
      missingIdentifiers.push(identifier)
    }
  }

  if (missingIdentifiers.length > 0) {
    return missingIdentifiers
  }

  return null
}

export function buildWhereUniqueInput(data: Record<string, any>, uniqueIdentifiers: string[]) {
  if (uniqueIdentifiers.length === 1) {
    return pickFromRecord(data, uniqueIdentifiers)
  }

  const compoundName = uniqueIdentifiers.join('_')

  return {
    [compoundName]: pickFromRecord(data, uniqueIdentifiers),
  }
}

function pickFromRecord(record: Record<string, any>, keys: string[]) {
  const output: Record<string, any> = {}

  for (const identifier of keys) {
    if (record[identifier]) {
      output[identifier] = record[identifier]
    }
  }

  return output
}
