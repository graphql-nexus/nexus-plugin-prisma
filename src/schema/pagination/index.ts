import { DMMF } from '@prisma/client/runtime'
import { DmmfTypes } from '../dmmf'
import { prismaStrategy } from './prisma'
import { relayStrategy } from './relay'

interface NormalizedPaginationArgs {
  cursor?: object
  skip?: string | number
  take?: string | number
  orderBy?: { [x: string]: 'asc' | 'desc' }
  [x: string]: any
}

export interface PaginationStrategy<GraphQLPaginationArgs extends object = object> {
  paginationArgNames: (keyof GraphQLPaginationArgs)[]
  transformDmmfArgs(params: {
    args: DmmfTypes.SchemaArg[]
    paginationArgNames: string[]
    field: DMMF.SchemaField
  }): DmmfTypes.SchemaArg[]
  resolve(args: GraphQLPaginationArgs): NormalizedPaginationArgs
}

// todo remove, but TS error if we do, dont' understand the error message
interface PaginationStrategies {
  relay: typeof relayStrategy
  prisma: typeof prismaStrategy
}

export const paginationStrategies: PaginationStrategies = {
  relay: relayStrategy,
  prisma: prismaStrategy,
}

export type PaginationStrategyTypes = typeof relayStrategy | typeof prismaStrategy
