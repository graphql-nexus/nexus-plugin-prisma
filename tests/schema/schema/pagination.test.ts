import { objectType } from '@nexus/schema'
import { paginationStrategies } from '../../../src/schema/pagination'
import { generateSchemaAndTypes } from '../__utils'

it('support relay pagination (default)', async () => {
  const datamodel = `
datasource db {
  provider = "postgresql"
  url      = "postgresql://"
}

model User {
  id Int @id @default(autoincrement())
  posts  Post[]
}

model Post {
  id Int @id @default(autoincrement())
  title String
  author  User @relation(fields: [authorId], references: [id])
  authorId Int
}
`

  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.posts({
        pagination: true,
      })
    },
  })

  const Post = objectType({
    name: 'Post',
    definition(t: any) {
      t.model.id()
      t.model.title()
    },
  })

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [User, Post])

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})

it('support prisma pagination', async () => {
  const datamodel = `
  datasource db {
    provider = "postgresql"
    url      = "postgresql://"
  }

  model User {
    id Int @id @default(autoincrement())
    posts  Post[]
  }

  model Post {
    id Int @id @default(autoincrement())
    title String
    author User @relation(fields: [authorId], references: [id])
    authorId Int
  }
`

  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.posts({
        pagination: true,
      })
    },
  })

  const Post = objectType({
    name: 'Post',
    definition(t: any) {
      t.model.id()
      t.model.title()
    },
  })

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [User, Post], {
    paginationType: 'prisma',
    paginationStrategy: paginationStrategies.prisma,
  })

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})
