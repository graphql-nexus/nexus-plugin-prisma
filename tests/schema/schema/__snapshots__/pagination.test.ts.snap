// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`support prisma pagination: schema 1`] = `
"type Post {
  id: Int!
  title: String!
}

input PostWhereUniqueInput {
  id: Int
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  posts(take: Int, skip: Int, cursor: PostWhereUniqueInput): [Post!]!
}
"
`;

exports[`support prisma pagination: typegen 1`] = `
"import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
  take?: boolean
  skip?: boolean
  cursor?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  User: Prisma.User
  Post: Prisma.Post
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    findManyUser: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'posts'
      ordering: 'id'
    }
    findManyPost: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'title' | 'author' | 'authorId'
      ordering: 'id' | 'title' | 'authorId'
    }
  },
  User: {
    posts: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'title' | 'author' | 'authorId'
      ordering: 'id' | 'title' | 'authorId'
    }
  }
  Post: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    findOneUser: 'User'
    findManyUser: 'User'
    findOnePost: 'Post'
    findManyPost: 'Post'
  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOnePost: 'Post'
    updateOnePost: 'Post'
    updateManyPost: 'BatchPayload'
    deleteOnePost: 'Post'
    deleteManyPost: 'BatchPayload'
    upsertOnePost: 'Post'
  },
  User: {
    id: 'Int'
    posts: 'Post'
  }
  Post: {
    id: 'Int'
    title: 'String'
    author: 'User'
    authorId: 'Int'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  User: Typegen.NexusPrismaFields<'User'>
  Post: Typegen.NexusPrismaFields<'Post'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;

exports[`support relay pagination (default): schema 1`] = `
"type Post {
  id: Int!
  title: String!
}

input PostWhereUniqueInput {
  id: Int
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  posts(first: Int, last: Int, before: PostWhereUniqueInput, after: PostWhereUniqueInput): [Post!]!
}
"
`;

exports[`support relay pagination (default): typegen 1`] = `
"import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
  first?: boolean
  last?: boolean
  before?: boolean
  after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  User: Prisma.User
  Post: Prisma.Post
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    findManyUser: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'posts'
      ordering: 'id'
    }
    findManyPost: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'title' | 'author' | 'authorId'
      ordering: 'id' | 'title' | 'authorId'
    }
  },
  User: {
    posts: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'title' | 'author' | 'authorId'
      ordering: 'id' | 'title' | 'authorId'
    }
  }
  Post: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    findOneUser: 'User'
    findManyUser: 'User'
    findOnePost: 'Post'
    findManyPost: 'Post'
  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOnePost: 'Post'
    updateOnePost: 'Post'
    updateManyPost: 'BatchPayload'
    deleteOnePost: 'Post'
    deleteManyPost: 'BatchPayload'
    upsertOnePost: 'Post'
  },
  User: {
    id: 'Int'
    posts: 'Post'
  }
  Post: {
    id: 'Int'
    title: 'String'
    author: 'User'
    authorId: 'Int'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  User: Typegen.NexusPrismaFields<'User'>
  Post: Typegen.NexusPrismaFields<'Post'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;
